### PR TITLE
Add a PhoneNumber class to handle formatting and actions on numbers

### PR DESCRIPTION
**Pull request by ryanbales**
We value collaboration at Invoca, and focus on peer mentorship as a way for the whole team to learn and thrive. As part of this process, we conduct code reviews during our development cycle. In this phase of the interview, we would like you to review and comment on this Pull Request. 

You can assume that this Pull Request was written by an intern, who is new to professional software development.  Please take about 20-30 minutes to review this Pull Request and comment as if you were reviewing an intern's code.

### CODE REVIEW

File: phone_number_spec.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+require_relative "./phone_number"
+
```
Comment (tbrussellCA) [inline phone_number_spec.rb:2]: You may also need to require the actions file.

File: phone_actions.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class PhoneActions
```
Comment (tbrussellCA) [inline phone_actions.rb:1]: Overall great job. Please add additional comments to the beginning of the files providing a bit more context as well as to some of the functions.

File: phone_number.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    PhoneActions.new(USERNAME, PASSWORD).text!(@number, message)
+  end
+
+  def call(message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(@number, message)
+  end
+
+  def text_and_call(message)
+    PhoneActions.new(USERNAME, PASSWORD).text!(@number, message)
+    PhoneActions.new(USERNAME, PASSWORD).call!(message, message)
```
Comment (tbrussellCA) [inline phone_number.rb:22]: should the first parameter here be @number?

File: phone_number.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def find_exchange
+    @number.gsub!(/[^0-9]/, "")
+    @number[3..5]
+  end
+
+  def find_line_number
+    @number.gsub!(/[^0-9]/, "")
+    @number[-4..@number.length]
```
Comment (tbrussellCA) [inline phone_number.rb:60]: a handy shortcut is -1 instead of @number.length to go the the end of the array

File: phone_actions.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+  def initialize(username, password)
+    @actions_url = "https://#{username}:#{password}@api.<made_up>.com/actions/"
+  end
+
+  def call!(number, message)
+    JSON.parse(RestClient.post("@actions_url/#{text}", {message: message, number: number}))
```
Comment (tbrussellCA) [inline phone_actions.rb:8]: Can you add additional logic to verify a successful status and prevent a failure from causing exceptions? You may want to add logging to more easily troubleshoot future issues.

File: phone_actions.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  end
+
+  def text!(number, message)
+    JSON.parse(RestClient.post("@actions_url/#{call}", {message: message, number: number}))
```
Comment (tbrussellCA) [inline phone_actions.rb:12]: Where are #{text} and #{call} being set?

File: phone_number.rb [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  private
+
+  def format_number
+    "(#{find_area_code}) #{find_exchange}-#{find_line_number}"
+  end
+
+  def find_area_code
+    open_paren = @number.index('(')
```
Comment (tbrussellCA) [inline phone_number.rb:41]: It looks like you handled the phone number parsing quite well. As an enhancement can you try refactoring this code so that it is a little more condensed? For example, you could combine the nested ifs, try if statement modifiers, or move the index adjustments to the section where you define open_paren & close-paren.

File: phone_number.rb [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        @number[(open_paren + 1)..(close_paren - 1)]
+      end
+    else
+      @number.gsub!(/[^0-9]/, "")
+      @number[0..2]
+    end
+  end
+
+  def find_exchange
+    @number.gsub!(/[^0-9]/, "")
```
Comment (tbrussellCA) [inline phone_number.rb:54]: It looks like the gsub code is reused a few places. It might be convenient to make this a function with optional parameters to subset the number so that it can be recycled for determining the area code, exchange, and line number.

File: phone_number_spec.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+RSpec.describe PhoneNumber do
+  let(:phone_number) { PhoneNumber.new("5558675309") }
```
Comment (tbrussellCA) [inline phone_number_spec.rb:4]: Please also include additional tests to see how it responds to nil/invalid phone numbers. Your original functions may also need corresponding adjustments.

**Review by tbrussellCA — COMMENTED** [review]
Great start. Made comments for a few potential issues/areas for improvement.
